### DISCUSSION TITLE
Craft Grid System

### DISCUSSION THREAD

**Discussion by angrybrad** [craftcms/cms#8103] [Ideas]
> Created by: Nathan Doyle (natetronn@gmail.com) on 2015/03/11 04:20:14 +0000
> Votes at time of UserVoice import: 97

-----------------------------------

I'd like to request a Grid System for all our content modeling and or layout needs. Such a thing will really complete the content modeling within Craft. 

Please see the following Gist for an in depth proposal which includes ideas, general brain storming and a very early stage mockup etc:

https://gist.github.com/Natetronn/7ee8ee1f8788702d06cb

Please keep comments here on the feedback site since this, I assume, is where P&T will get notifications about said feedback (or at least include a link to comment here as well.)

Note: I'm adding this as a core FR. Not because there isn't any interest in building it from the plugin dev community, rather, because the amount of UI/UX which would be required to go into something like this, the amount of use it would get (I'm sure a ton!) and the fact that any change in UI/CP on the part of P&T would surely require updates to such a plugin on a constant basis making it ripe for UI bugs and general outcry ;-) 

I also realize everyone wants all the things in core and I don't suggest doing such a thing lightly of course, especially something of this size but, I think this is one of those that would at the very least need P&T input from the start. That way who ever was to build was not going it alone, blindly. Building on a strong base etc.

Thanks for you consideration!

**Comment by angrybrad** [comment]
> Posted by Nathan Doyle (natetronn@gmail.com) on 2015/03/11 04:20:14 +0000
-----------------------------------

This is interesting: http://gridster.net

**Comment by angrybrad** [comment]
> Posted by Nathan Doyle (natetronn@gmail.com) on 2015/03/11 04:20:14 +0000
-----------------------------------

@PlutoMars I promoted this a few times it&#39;s true! And I definitely added a ton of &quot;promotional materials&quot; to get my ideas and points across to everyone the best I could.

I&#39;m not sure it&#39;s a distorted view I created myself, however. I don&#39;t hold a gun to people&#39;s heads and make them vote. Such a thing being said kind of bothers me in fact because it implies people around her don&#39;t think for themselves and that of course is so very far from the truth.

I am also in favor of seeing Matrix within Matrix by the way.

**Comment by angrybrad** [comment]
> Posted by PlutoMars (Limey.pluto@gmail.com) on 2015/03/20 09:37:19 +0000
-----------------------------------

@Nathan, let&#39;s also not forget you advertised this plugin as well too alot :), none of the other feature requests got promoted like that. so writing that it&#39;s more popular then &quot;matrix in matrix&quot; is a distorted view you created yourself...

**Comment by angrybrad** [comment]
> Posted by Nathan Doyle (natetronn@gmail.com) on 2015/03/11 04:20:14 +0000
-----------------------------------

@K Moens. Thanks for the feedback!

You might see it most of the time in WP because there are many similar options that exist for it today. And of course WP&#39;s popularity etc. In fact, I listed a few in the gist as context. I was worried the association would taint said context, however.

I listed only a few but, there are probably 20+ others which I didn&#39;t list for WP alone.

Something for layout exists either natively or via plugins in almost all the major CMS these days. That includes Drupal, Joomla, WP, Concrete5, and even Statamic is releasing something soon as well.

This is a very popular CMS feature to say the least.

With that said, I&#39;m all for working on a solution which is better than anything which currently exists. Something that is mobile friendly, that is inline with the Craft paradigm and takes Grid Layouts, as well as Long Form Content, into consideration; all while being user friendly for both content editors and developers alike!

This idea is more popular than Matrix within Matrix and is currently resting in at the 7th place out of the top 10 requests. That&#39;s allot of interest if you ask me and I&#39;m really hoping, if this does get approved, that there is more feedback on how to do it right should P&amp;T require it.

Then again, this isn&#39;t a democracy and it&#39;s up to P&amp;T to decide what gets included or not so this could all be for not ;-)

**Comment by angrybrad** [comment]
> Posted by K Moens (kurtmoens@gmail.com) on 2015/05/18 18:34:40 +0000
-----------------------------------

Personally don&#39;t need this... too wordpressy. Certainly not as 1st party product.

**Comment by angrybrad** [comment]
> Posted by Nathan Doyle (natetronn@gmail.com) on 2015/03/11 04:20:14 +0000
-----------------------------------

Concept two of the Craft Grid System is now known as the &quot;Layout Manager&quot;:

https://gist.github.com/Natetronn/7ee8ee1f8788702d06cb#comment-1453780

This is very rough but, I hope the idea comes across just the same. Feedback always welcome!

**Comment by angrybrad** [comment]
> Posted by Nathan Doyle (natetronn@gmail.com) on 2015/03/11 04:20:14 +0000
-----------------------------------

@sean you&#39;re right I did! I said in the gist either &quot;core or plugin&quot; but, only because I didn&#39;t know what would be possible or not. This is just an idea of course. The more I gave it thought the more I thought it would work as a plugin, however. You know, similar to how Matrix or Lightswitch is part of core in that they are included with Craft itself but, at the same time exist as plugins. In other words, first party. 

Does that make sense?

And by the way, since beggars can&#39;t be choosers, I&#39;m all for 3rd party devs tackling this if they feel the urge to do so and or if P&amp;T doesn&#39;t add it to their list. I will say that I&#39;ve seen a pretty good 3rd party solution for the Matrix within Matrix request via SuperTable within Matrix but, people still were asking for a 1st party solution instead so I felt it best to start with the source first.

@Narration SD yes, that&#39;s an idea. Similar to how plugins can tie itself to the system? I&#39;m all for discussing such a thing in a separate request. In fact, I had even grander ideas for Craft Grid System but, I held back in fears that we&#39;d end up with nada versus algo ;-)

**Comment by angrybrad** [comment]
> Posted by Narration SD (narrationsd@gmail.com) on 2015/04/14 16:07:09 +0000
-----------------------------------

After hoping this will stay out of the #announcements on Slack now (!) found a thought coming about it.

It seems it could be very helpful to turn this topic into what would be a kind of CP Abstraction Layer. 

This would allow and encourage building any specialist UI like this on top of the very well thought out general capabilities of Craft -- letting them then provide the end user targeting simplifications for given scenarios that actually could use the help.

This would include rectangular boxes of boxes as well as interesting challenges like the matrix driven auto-layout articles at the heart of Craft&#39;s initiating intentions. That interview yesterday with Brandon underlined a lot.

I would think such a layer could bring those article tools full circle, letting the resulting fully visible article be front and center, with a toolpad on the side to compactly add elements -- and let me say no more.

I can imagine the CP Abstraction Layer giving us tools for quickly arranging all manner of custom CP blocks as well, keeping Craft in control while allowing all the simplifying things for customers.

One particular matter that it would answer for me personally is how neat and effective we can appreciate the auto layout Matrix articles are, and how much capable agency customers are vocal about liking this, but knowing they are at moment nearly a bridge too far still for more individual and cms-naive clients: the art galleries, doctors offices, institutions, etc..

Two centavos, anyway...

( I&#39;m near a border, and enjoying the conversations around the house)

**Comment by angrybrad** [comment]
> Posted by Sean O&#39;Grady (sean@miralize.com) on 2015/03/19 11:14:36 +0000
-----------------------------------

If thats you mean then yep I agree! But above you say &quot;I&#39;m adding this as a core FR. &quot;

**Comment by angrybrad** [comment]
> Posted by Nathan Doyle (natetronn@gmail.com) on 2015/03/11 04:20:14 +0000
-----------------------------------

@sean and just to be clear, the idea, in theory, is actually for it to live as a plugin (or three maybe - Section, Row, Column which each require the other)

I just put it into the content modeling section because it seemed to fit there best.

**Comment by yellostudio** [comment]
Almost every site requires some level of layout control and despite the flexibility of crafts matrix, its not the most intuitive for content layout.  It looks as though this thread is old.  When working with the CMS currently this is just a complete nightmare for anything complex.  Wordpress whether you like or loathe it have understood this side of content management.  Having a layout builder before adding content or components would make Craft a weapon.  

There have been attempts to achieve this in the past;
Little Layout
Vizzy plugin ](https://verbb.io/craft-plugins/vizy/features) 

But in my opinion it should in core and an option/setting to be enabled.

I have been running an agency for many years and we have built literally everything in craft from the simplest to most complex of sites and systems.  

This part of craft has always let our end product down.

**Reply by thupsi** [reply]
While I'm not a fan of page builders in general (I dislike mixing content with presentation, or giving styling options to authors), I recognize that some basic level of layout tooling can be useful in one-off landing pages etc, and in any situation where proper content architecture, reusability and structure is not the first priority. 

An implementation that seems to hit the right balance is the [layout field](https://getkirby.com/docs/reference/panel/fields/layout) in Kirby.

**Reply by alexanderbuergin** [reply]
I agree with this. I believe Craft CMS should implement this option as a core field type, as described below.

**Comment by alexanderbuergin** [comment]
While I’m also not the biggest fan of content-focused page builders, I recognize their value, particularly for creating more complex, marketing-driven landing pages that are only intended for one-off or occasional use. Having a content or column-based layout builder would be a substantial enhancement for Craft CMS, especially for cases where it’s not sustainable to create a new module every time a custom layout is needed.

We recently took over a site from another agency that had over 150 Matrix blocks, making it nearly impossible to manage or even understand the structure. A well-integrated layout solution could alleviate this kind of clutter and improve maintainability.

Ideally, this feature could work as a field type that allows us to embed Matrix fields within it or even use it as a field type inside existing Matrix blocks. This would let us retain the flexibility of Matrix while adding a content layout builder directly within the workflow. For instance, it could allow users to create a Matrix block with a specific field type that organizes three or four columns, each with a Matrix or CKEditor field. Ideally, both options would be available, allowing endless content combinations in a structured, logical way.

In my view, implementing this as a standalone field type would be the most effective approach. It would allow seamless integration with existing workflows and configurations through settings within the field itself, making it adaptable to any layout without disrupting Craft CMS’s current functionality.

In my opinion, this feature should ideally be part of the core, with an option to enable it as needed.

**Reply by yellostudio** [reply]
[alexanderbuergin](https://github.com/alexanderbuergin) this is what i had in mind too

**Reply by wsydney76** [reply]
> An implementation that seems to hit the right balance is the [layout field](https://getkirby.com/docs/reference/panel/fields/layout) in Kirby.

Agree, there is an old FR for this: #8998

Conceptually, this is 'just' a variation of a blocks/matrix field, where blocks/nested entries are arranged in a grid, instead of in a linear sequence. So it could be used everywhere where a matrix/CKEditor field can be used, and can contain any nested content itself.

With all the work done for matrix/CKEditor, the content storage side of things shouldn't be that hard to implement.

Think of cards/inline editable blocks in this example:

![image](https://github.com/user-attachments/assets/94c25ba4-0156-4687-b9c8-9dbb63f4ffd1)

**Reply by alexanderbuergin** [reply]
Would love to see something like that.

**Comment by piotrpog** [comment]
There was imperavi library called "article" that allowed for creation of multi-column layout. Now it seems that it was discontinued, and its functionalities included into redactor. But im not sure if it is same redactor version that craft cms redactor plugin uses.

**Reply by alexanderbuergin** [reply]
Redactor is no longer part of the core support for Craft CMS.

Tt would be better not to handle this within an RTF field. A native field would be ideal to keep the data clean and well-structured. 😊

**Comment by humandevs** [comment]
Kind of like how CSS avoided a "grid" or "tables" feature for years and years for purity, and then finally caved and now CSS Grid is used everywhere, Componentized grids are almost everywhere these days, especially in front end frameworks like React.

Craft actually has this concept in the editor, where fields can be one to four columns wide.

+1 for the Divi builder in Wordpress, it really is a game-changer.

In a bigger picture, enabling a Grid System for dragging in components (Entry Types) into that grid and being able to toggle the component sized based on the sizes supported by that component enables lots of no-code complex custom application design.

CSS Grid also solves the issues of wrapping grid divs when needed.

**Reply by alexanderbuergin** [reply]
Some thing like a Field Type
